CONTRIBUTING: Add a more fleshed out contributing guide (#4)
rapid7/insightconnect-plugins: CONTRIBUTING.md

# Contributing

> Thank you for your interest in improving InsightConnect! Please review our [Code of Conduct] before making contributions.

Reviewer: Thank you for your interest in joining the InsightConnect Developer community!

Author: Nice!

Author: Updated

# Contributing

Thank you for your interest in improving InsightConnect! Please review our [Code of Conduct] before making contributions.

There are multiple ways to contribute beyond writing code. These include:

- [Submit bugs and feature requests] with detailed information about your issue or idea.
- [Help fellow users with open issues] or [help fellow committers test recent pull requests].
- [Report a security vulnerability in InsightConnect's plugins] to Rapid7.
- Submit an updated or brand new plugin!  We are always eager for new
  integrations or features. Don't know where to start? Check out the [developer documentation].

Here is a short list of do's and don'ts to make sure *your* valuable contributions actually make

> it into production.  If you do not care to follow these rules, your contribution

Reviewer 2: ```suggestion
it into production.  If you do not care to follow these rules, your contribution will be rejected
```
Using `closed` implies a pull request is used. Might want to play around with the wording.

**will** be closed. Sorry!

## Code Contributions

- **Do** read the [developer documentation]
- **Do** stick to the [Python PEP8] and [plugin style] guides.
- **Do** follow the [50/72 rule] for Git commit messages.
- **Do** license your code as MIT.
- **Do** create a [topic branch] to work on. This helps ensure users are aware of commits on the branch being considered for merge, allows for a location for more commits to be offered without mingling with other contributor changes, and allows contributors to make progress while a PR is still being reviewed.

### Pull Requests

> - **Do** write "WIP" on your PR and/or open a [draft PR] if submitting **working** yet unfinished code.

Reviewer 2: Why does the code have to be working? I would omit that - WIP is totally valid for non-functional, but unfinished code.

Reviewer 2: Example: someone just wants eyes on it or is requesting assistance.

Author: Updated

- **Do** target your pull request to the **master branch**.
- **Do** specify a descriptive title to make searching for your pull request easier e.g. "Okta: add Suspend User action".
- **Do** include [console output], especially the JSON output for new features and bug fixes.
- **Do** list [verification steps] so your tests are reproducible.
- **Do** [reference associated issues] in your pull request description.
- **Don't** leave your pull request description blank.
- **Don't** abandon your pull request. Being responsive helps us land your code faster.

#### New Features

- **Do** install validator dependencies necessary to run `make validate` to find and fix any errors or warnings that come up.
- **Don't** include more than one plugin per pull request.

> - **Do** include documentation showing sample run-throughs.

Reviewer 2: Can you put all the `do's` above the `don'ts`?

Author: Reordered

# Contributing

> Thanks for your interest in making InsightConnect a better security orchestration tool.

Reviewer 2: ```suggestion
Thank you for your interest in making InsightConnect a better security orchestration tool. Please review our [Code of Conduct] before making contributions.
```
More formal

> Before you get started, review our [Code of Conduct].

Reviewer 2: ```suggestion
```

> There are multiple ways to help beyond just writing code:

Reviewer 2: ```suggestion
There are multiple ways to contribute beyond writing code. These include:
```

- [Submit bugs and feature requests] with detailed information about your issue or idea.
- [Help fellow users with open issues] or [help fellow committers test recent pull requests].

> - [Report a security vulnerability in InsightConnect itself or its plugins] to Rapid7.

Reviewer 2: The scope of this repository is only plugins - I'm not sure I would mention the InsightConnect product

Author: Done

- Submit an updated or brand new plugin!  We are always eager for new
  integrations or features. Don't know where to start? Check out the [developer documentation].

> Here's a short list of do's and don'ts to make sure *your* valuable contributions actually make

Reviewer 2: ```suggestion
Here is a short list of do's and don'ts to make sure *your* valuable contributions actually make
```

> it into the master branch.  If you do not care to follow these rules, your contribution

Reviewer 2: ```suggestion
it into production.  If you do not care to follow these rules, your contribution
```

**will** be closed. Sorry!

## Code Contributions

- **Do** read the [developer documentation]
- **Do** stick to the [Python PEP8] and [plugin style] guides.
- **Do** follow the [50/72 rule] for Git commit messages.
- **Do** license your code as MIT.

> - **Do** create a [topic branch] to work on instead of working directly on `master`. This helps protect the process, ensures users are aware of commits on the branch being considered for merge, allows for a location for more commits to be offered without mingling with other contributor changes, and allows contributors to make progress while a PR is still being reviewed.

Reviewer 2: I'm not sure the first part of this statement is necessary. We *should* have merge protection enabled on `master` making it impossible to merge into it without going through a PR process which requires a branch.

Author: Done

### Pull Requests

- **Do** write "WIP" on your PR and/or open a [draft PR] if submitting **working** yet unfinished code.
- **Do** target your pull request to the **master branch**.

> - **Do** specify a descriptive title to make searching for your pull request easier.

Reviewer 2: How about requiring a format for this? Ie. "Plugin Name - Brief description of change"

Author: Good idea

Author: Done

- **Do** include [console output], especially the JSON output for new features and bug fixes.

> - **Do** list [verification steps] so your code is testable.

Reviewer 2: ```suggestion
- **Do** list [verification steps] so your tests are reproducible.
```
?

- **Do** [reference associated issues] in your pull request description.
- **Don't** leave your pull request description blank.

> - **Don't** abandon your pull request. Being responsive helps us land your code faster.

Reviewer 2: Would there be any benefit to cleaning out inactive pull requests after x days?

Author: We can think about this

#### New Features

- **Do** install validator dependencies necessary to run `make validate` to find and fix any errors or warnings that come up.
- **Don't** include more than one plugin per pull request.
- **Do** include documentation showing sample run-throughs.

#### Bug Fixes

- **Do** include reproduction steps in the form of [verification steps].
- **Do** link to any corresponding [Issues] in the format of `See #1234` in your commit description.

## Bug Reports

Please report vulnerabilities in Rapid7 software directly to security@rapid7.com.
For more on our disclosure policy and Rapid7's approach to coordinated disclosure, [head over here](https://www.rapid7.com/security).

When reporting issues:

- **Do** write a detailed description of your bug and use a descriptive title.
- **Do** include reproduction steps, stack traces, and anything that might help us fix your bug.
- **Don't** file duplicate reports; search for your bug before filing a new report.

> If you need some more guidance, talk to the main body of open source contributors over at

Reviewer 2: ```suggestion
If you need additional guidance, reach out to the open source contribution owners at
```
